octobot_commons/async_job.py

```diff
@@ -0,0 +1,127 @@
+#
+#  You should have received a copy of the GNU Lesser General Public
+#  License along with this library.
+import asyncio
+
+import time
+
+from octobot_commons.asyncio_tools import wait_asyncio_next_cycle
+from octobot_commons.logging.logging_util import get_logger
+
+
+class AsyncJob:
+    """
+    Async job management
+    """
+
+    NO_DELAY = -1
+
+    def __init__(
+        self,
+        callback,
+        execution_interval_delay=NO_DELAY,
+        min_execution_delay=NO_DELAY,
+        job_dependencies=None,
+    ):
+        self.logger = get_logger(self.__class__.__name__)
+        self.is_running = False
+        self.is_scheduled = False
+        self.last_execution_time = 0
+        self.execution_interval_delay = execution_interval_delay
+        self.min_execution_delay = min_execution_delay
+        self.callback = callback
+        self.job_dependencies = job_dependencies if job_dependencies else []
+        self.job_task = None
+
+    async def run(self):
+        """
+        Run the job if possible
+        Reschedule the jab in the end
+        """
+        if self._should_run():
+            await self._run()
+        if not self.is_scheduled:
+            await self._reschedule()
+
+    def is_job_running(self):
+        """
+        :return: publicly is_running attribute value
+        """
+        return self.is_running
+
+    async def _run(self):
+        """
+        Execute the job callback
+        Reset the last_execution_time
+        """
+        self.is_running = True
+        try:
+            await self.callback()
+        except Exception as exception:
+            self.logger.error(f"Failed to run job action : {exception}")
+        finally:
+            self.last_execution_time = time.time()
+            self.is_running = False
+            self.is_scheduled = False
+
+    async def _reschedule(self):
+        """
+        Reschedule the job according to execution_interval_delay
+        if execution_interval_delay == AsyncJob.NO_DELAY, tries to execute without delay (in the next async loop)
+        else schedule the execution at now + execution_interval_delay
+        """
+        if self.execution_interval_delay != AsyncJob.NO_DELAY:
+            self.job_task = asyncio.create_task(self._postpone_run())
+        else:
+            await wait_asyncio_next_cycle()
+            await self.run()
+        self.is_scheduled = True
+
+    async def _postpone_run(self):
+        """
+        Postpone the run() call at execution_interval_delay
+        """
+        await asyncio.sleep(self.execution_interval_delay)
+        await self.run()
+
+    def _should_run(self):
+        """
+        Return True if the job is not already running, if dependent jobs are also not running
+        and if min_execution_delay < last_execution_time
+        """
+        return (
+            not self.is_running
+            and not any([job.is_running for job in self.job_dependencies])
```

Comment:
that will simply cancel a call on running dependencies ? I think we should wait for them to finish instead of cancelling the whole call

Reply:
Good idea, we shoud allow this :+1: